File: packages/expo-speech/android/src/main/java/expo/modules/speech/SpeechModule.java
package expo.modules.speech;

import android.content.Context;
import android.os.Bundle;
import android.speech.tts.TextToSpeech;
import android.speech.tts.UtteranceProgressListener;
import android.speech.tts.Voice;

import org.unimodules.core.ExportedModule;
import org.unimodules.core.ModuleRegistry;
import org.unimodules.core.Promise;
import org.unimodules.core.interfaces.ExpoMethod;
import org.unimodules.core.interfaces.LifecycleEventListener;
import org.unimodules.core.interfaces.services.EventEmitter;
import org.unimodules.core.interfaces.services.UIManager;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Queue;

public class SpeechModule extends ExportedModule implements LifecycleEventListener {

  private TextToSpeech mTextToSpeech;
  private boolean mTtsReady = false;
  private Queue<Map<String, Object>> mDelayedUtterances = new ArrayDeque<>();
  private Context mContext;
  private ModuleRegistry mModuleRegistry;

  public SpeechModule(Context context) {
    super(context);
    mContext = context;
  }

  @Override
  public String getName() {
    return "ExponentSpeech";
  }

  @ExpoMethod
  public void speak(final String id, final String text, final Map<String, Object> options, final Promise promise) {
    if(text.length() > TextToSpeech.getMaxSpeechInputLength()) {
      promise.reject("ERR_SPEECH_INPUT_LENGTH",
        "Speech input text is too long! Limit of input length is: " + TextToSpeech.getMaxSpeechInputLength());
      return;
    }

    if (mTtsReady) {
      speakOut(id, text, options);
    } else {
      mDelayedUtterances.add(Collections.unmodifiableMap(new HashMap<String, Object>() {
        {
          put("id", id);
          put("text", text);
          put("options", options);
        }
      }));
      // init TTS, speaking will be available only after onInit
      getTextToSpeech();
    }
    promise.resolve(null);
  }

  private void speakOut(final String id, final String text, final Map<String, Object> options) {
    TextToSpeech textToSpeech = getTextToSpeech();
    if (options.containsKey("language")) {
      Locale locale = new Locale((String) options.get("language"));
      int languageAvailable = textToSpeech.isLanguageAvailable(locale);
      if (languageAvailable != TextToSpeech.LANG_MISSING_DATA &&
          languageAvailable != TextToSpeech.LANG_NOT_SUPPORTED) {
        textToSpeech.setLanguage(locale);
      } else {
        textToSpeech.setLanguage(Locale.getDefault());
      }
    } else {
      textToSpeech.setLanguage(Locale.getDefault());
    }
    if (options.containsKey("pitch")) {
      textToSpeech.setPitch(((Number) options.get("pitch")).floatValue());
    }
    if (options.containsKey("rate")) {
      textToSpeech.setSpeechRate(((Number) options.get("rate")).floatValue());
    }
    if (options.containsKey("voice")) {
      for (Voice voice : textToSpeech.getVoices()) {
        if (voice.getName().equals(options.get("voice"))) {
          textToSpeech.setVoice(voice);

          break;
        }
      }
    }

    textToSpeech.speak(
        text,
        TextToSpeech.QUEUE_ADD,
        null,
        id
    );
  }

  @ExpoMethod
  public void isSpeaking(Promise promise) {
    promise.resolve(getTextToSpeech().isSpeaking());
  }

  @ExpoMethod
  public void getVoices(Promise promise) {
    TextToSpeech textToSpeech = getTextToSpeech();
    ArrayList<Voice> nativeVoices = new ArrayList();
    ArrayList<Bundle> voices = new ArrayList();

    try {
      nativeVoices = new ArrayList<>(textToSpeech.getVoices());
    } catch (Exception e) {}

    for (Voice voice : nativeVoices) {
        Bundle voiceMap = new Bundle();

        String quality = "Default";

        if (voice.getQuality() > Voice.QUALITY_NORMAL) {
          quality = "Enhanced";
        }

        voiceMap.putString("identifier", voice.getName());
        voiceMap.putString("name", voice.getName());
        voiceMap.putString("quality", quality);
        voiceMap.putString("language", LanguageUtils.getISOCode(voice.getLocale()));

        voices.add(voiceMap);
    }

    promise.resolve(voices);
  }

  @ExpoMethod
  public void stop(final Promise promise) {
    getTextToSpeech().stop();
    promise.resolve(null);
  }

  private TextToSpeech getTextToSpeech() {
    if (mTextToSpeech == null) {
      mTextToSpeech = new TextToSpeech(mContext, new TextToSpeech.OnInitListener() {
        @Override
        public void onInit(int status) {
          if (status == TextToSpeech.SUCCESS) {
            // synchronize because in some cases this runs on another thread and mTTS is null
            synchronized (SpeechModule.this) {
              mTtsReady = true;
              mTextToSpeech.setOnUtteranceProgressListener(new UtteranceProgressListener() {
                @Override
                public void onStart(String utteranceId) {
                  EventEmitter emitter = mModuleRegistry.getModule(EventEmitter.class);
                  emitter.emit("Exponent.speakingStarted", idToMap(utteranceId));
                }

                @Override
                public void onDone(String utteranceId) {
                  EventEmitter emitter = mModuleRegistry.getModule(EventEmitter.class);
                  emitter.emit("Exponent.speakingDone", idToMap(utteranceId));
                }

                @Override
                public void onStop(String utteranceId, boolean interrupted) {
                  EventEmitter emitter = mModuleRegistry.getModule(EventEmitter.class);
                  emitter.emit("Exponent.speakingStopped", idToMap(utteranceId));
                }

                @Override
                public void onError(String utteranceId) {
                  EventEmitter emitter = mModuleRegistry.getModule(EventEmitter.class);
                  emitter.emit("Exponent.speakingError", idToMap(utteranceId));
                }
              });

              for (Map<String, Object> arguments : mDelayedUtterances) {
                speakOut(
                  (String) arguments.get("id"),
                  (String) arguments.get("text"),
                  (Map<String, Object>) arguments.get("options")
                );
              }
            }
          }
        }
      });
    }
    return mTextToSpeech;
  }

  @Override
  public Map<String, Object> getConstants() {
    final Map<String, Object> constants = new HashMap<>();
    constants.put("maxSpeechInputLength", TextToSpeech.getMaxSpeechInputLength());
    return constants;
  }

  @Override
  public void onHostResume() { }

  @Override
  public void onHostPause() { }

  @Override
  public void onHostDestroy() {
    getTextToSpeech().shutdown();
  }

  private Bundle idToMap(String id) {
    Bundle map = new Bundle();
    map.putString("id", id);
    return map;
  }

  @Override
  public void onCreate(ModuleRegistry moduleRegistry) {
    mModuleRegistry = moduleRegistry;

    // Register to new UIManager
    if (mModuleRegistry != null && mModuleRegistry.getModule(UIManager.class) != null) {
      mModuleRegistry.getModule(UIManager.class).registerLifecycleEventListener(this);
    }
  }

  @Override
  public void onDestroy() {
    // Unregister from old UIManager
    if (mModuleRegistry != null && mModuleRegistry.getModule(UIManager.class) != null) {
      mModuleRegistry.getModule(UIManager.class).unregisterLifecycleEventListener(this);
    }
    mModuleRegistry = null;
  }
}

File: packages/expo-speech/android/src/main/java/expo/modules/speech/SpeechModule.kt
package expo.modules.speech

import android.os.Bundle
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import android.speech.tts.Voice
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.ArrayDeque
import java.util.Locale
import java.util.Queue

const val speakingStartedEvent = "Exponent.speakingStarted"
const val speakingWillSayNextStringEvent = "Exponent.speakingWillSayNextString"
const val speakingDoneEvent = "Exponent.speakingDone"
const val speakingStoppedEvent = "Exponent.speakingStopped"
const val speakingErrorEvent = "Exponent.speakingError"

class SpeechModule : Module() {
  private val delayedUtterances: Queue<Utterance> = ArrayDeque()

  override fun definition() = ModuleDefinition {
    Name("ExpoSpeech")

    Events(
      speakingStartedEvent,
      speakingWillSayNextStringEvent,
      speakingDoneEvent,
      speakingStoppedEvent,
      speakingErrorEvent
    )

    Constants("maxSpeechInputLength" to TextToSpeech.getMaxSpeechInputLength())

    OnActivityDestroys {
      textToSpeech.shutdown()
    }

    AsyncFunction<Boolean>("isSpeaking") {
      textToSpeech.isSpeaking
    }

    AsyncFunction<List<VoiceRecord>>("getVoices") {
      val nativeVoices = try {
        textToSpeech.voices.toList()
      } catch (_: Exception) {
        emptyList()
      }

      return@AsyncFunction nativeVoices.map {
        val quality = if (it.quality > Voice.QUALITY_NORMAL) {
          VoiceQuality.ENHANCED
        } else {
          VoiceQuality.DEFAULT
        }

        VoiceRecord(
          identifier = it.name,
          name = it.name,
          quality = quality,
          language = LanguageUtils.getISOCode(it.locale)
        )
      }
    }

    AsyncFunction<Unit>("stop") {
      textToSpeech.stop()
    }

    AsyncFunction("speak") { id: String, text: String, options: SpeechOptions ->
      if (text.length > TextToSpeech.getMaxSpeechInputLength()) {
        throw SpeechInputIsToLongException()
      }

      if (isTextToSpeechReady) {
        speakOut(id, text, options)
      } else {
        delayedUtterances.add(Utterance(id, text, options))

        // init TTS, speaking will be available only after onInit
        textToSpeech
      }
      Unit
    }
  }

  private fun speakOut(id: String, text: String, options: SpeechOptions) {
    options.pitch?.let(textToSpeech::setPitch)
    options.rate?.let(textToSpeech::setSpeechRate)

    textToSpeech.language = options.language?.let {
      val locale = Locale(it)
      val languageAvailable = textToSpeech.isLanguageAvailable(locale)

      return@let if (
        languageAvailable != TextToSpeech.LANG_MISSING_DATA &&
        languageAvailable != TextToSpeech.LANG_NOT_SUPPORTED
      ) {
        locale
      } else {
        Locale.getDefault()
      }
    } ?: Locale.getDefault()

    options.voice?.let { voiceName ->
      textToSpeech.voices
        .firstOrNull { it.name == voiceName }
        ?.let(textToSpeech::setVoice)
    }

    textToSpeech.speak(
      text,
      TextToSpeech.QUEUE_ADD,
      null,
      id
    )
  }

  // TextToSpeech object related code

  private val isTextToSpeechReady
    get() = _ttsReady

  private val textToSpeech: TextToSpeech by lazy {
    val newTtsInstance = TextToSpeech(appContext.reactContext) { status: Int ->
      if (status == TextToSpeech.SUCCESS) {
        // synchronize because in some cases this runs on another thread and _textToSpeech is null
        synchronized(this@SpeechModule) {
          _ttsReady = true
          _textToSpeech!!.setOnUtteranceProgressListener(object : UtteranceProgressListener() {

            override fun onStart(utteranceId: String) {
              sendEvent(speakingStartedEvent, idToMap(utteranceId))
            }

            override fun onRangeStart(utteranceId: String, start: Int, end: Int, frame: Int) {
              val map = Bundle().apply {
                putString("id", utteranceId)
                putInt("charIndex", start)
                putInt("charLength", end - start)
              }
              sendEvent(speakingWillSayNextStringEvent, map)
            }

            override fun onDone(utteranceId: String) {
              sendEvent(speakingDoneEvent, idToMap(utteranceId))
            }

            override fun onStop(utteranceId: String, interrupted: Boolean) {
              sendEvent(speakingStoppedEvent, idToMap(utteranceId))
            }

            override fun onError(utteranceId: String) {
              sendEvent(speakingErrorEvent, idToMap(utteranceId))
            }
          })
          for ((id, text, options) in delayedUtterances) {
            speakOut(id, text, options)
          }
        }
      }
    }
    _textToSpeech = newTtsInstance
    newTtsInstance
  }

  // Helpers
  private fun idToMap(id: String) = Bundle().apply {
    putString("id", id)
  }

  private data class Utterance(
    val id: String,
    val text: String,
    val options: SpeechOptions
  )

  // do not refer to these - they're only needed when initializing `textToSpeech`
  private var _textToSpeech: TextToSpeech? = null
  private var _ttsReady = false
}
